docs/src/v1-api/consuming/override.md

```diff
@@ -0,0 +1,198 @@
+# Consume and modify dream2nix packages
+
+## Given the following package
+`upstream/my-package.nix`
+```nix
+{config, lib, dream2nix, ...}: {
+
+  imports = [
+    dream2nix.modules.nodejs.mkDerivation
+    dream2nix.modules.nodejs.package-lock
+  ];
+
+  pname = "my-package";
+  version = "2.0.0";
+
+  src = {
+    type = github;
+    owner = "my-user";
+    repo = "my-repo";
+    ref = config.version;
+    hash = "sha256-mia90VYv/YTdWNhKpvwvFW9RfbXZJSWhJ+yva6EnLE8=";
+  };
+
+  # declare dependency on python3
+  deps = {nixpkgs, ...}: {
+    python3 = nixpkgs.python39;
+  };
+
+  nativeBuildInputs = [
+    config.deps.python3
+  ];
+
+  configurePhase = ''
+    python3 --version
+  '';
+
+  buildPhase = ''
+    python3 -c 'print("Hello World!")' > $out
+  '';
+}
+```
+
+`upstream/default.nix`
+```nix
+{
+  nixpkgs ? import <nixpkgs> {},
+  dream2nix ?
+    import
+    (builtins.fetchTarball "https://dream2nix.dev/tarball/1.0")
+    {inherit nixpkgs;},
+}: {
+  packages.my-package = dream2nix.eval ./my-package.nix;
+}
+```
+
+## 1. Override using modules
+
+### 1.1 Define a module for the override
+`my-package-override.nix`
+```nix
+{config, lib, ... }: {
+
+  version = "2.1.0";
+
+  # No need to re-define other fetcher attributes.
+  # The module system updates them for us.
+  src.hash = "sha256-LM5GDNjLcmgZVQEeANWAOO09KppwGaYEzJBjYmuSwys=";
+
+  deps = {nixpkgs, ...}: {
```

Comment:
it's not clear why deps take another function. couldn't nixpkgs be passed to the parent function?

Reply:
> it's not clear why deps take another function

`deps` is our equivalent of package function args in nixpkgs. A single place to override dependencies.

> couldn't nixpkgs be passed to the parent function?

We could pass nixpkgs via `specialArgs` but that would incurage usage like:
```
{config, nixpkgs, ...}: {
  buildInputs = [nixpkgs.python3];
  nativeBuildInputs = [nixpkgs.python3];
  preConfigure = ''
    ${nixpkgs.python3}/bin/python --version
  '';
}
```
This would make it hard to override `python3`, as there isn't a single source of truth.

Instead we want to limit the scope of `packageSets` like `nixpkgs` to a single attribute `deps` which makes overriding simple.

I should add some comments to the code to explain more about it.

Another cool thing is that`deps` is a submodule, so it allows us to specify options:
```nix
{config, lib, ...}: {
  options.deps.python3 = lib.mkOption {
    description = "Select a python3 version for building this package";
    # ...
  };
}
```

... which allows us to generate a manual for each individual package.